### PR TITLE
feat : pull count calculator 2147

### LINKED ISSUE #2147: 新功能建议：抽数计算
### 功能描述

ocr识别，根据当前嵌晶玉数量、(源石数量-29)*75折算嵌晶玉数量、可兑换抽数数量判断当前用户可拥有多少限定抽，方便用户规划抽卡资源以及无井不下池，打印在程序界面，拒绝鹰角骗下池骗斩杀骗氪小巧思！

### 日志文件

无

### MaaEnd 版本

_No response_

### 其他信息

_No response_

### PR DESCRIPTION
**Pull request by qiongtu2077**
Closes #2147 

feat: 添加抽数计算任务 (#2147)

新增 PullCountCalculator 任务，用于读取寻访资源、扫描仓库寻访券，并输出当前池/下版本池子抽数。

- 从干员寻访页读取嵌晶玉与衍质源石换算嵌晶玉

- 自动进入贵重品库 > 珍贵物品扫描寻访券

- 按当前池限时券、可继承券、下版本专用券分类统计

- 加入下版本商店 5 抽和签到 5 抽

- 增加仓库滚动到底/重复页检测，避免循环扫描

- 补充任务入口、Pipeline、券配置、多语言文案和 Go 单测

Tests: go test ./pullcount; go test ./...; python tools/build_and_install.py

## Summary by Sourcery

添加一个 PullCountCalculator 自定义动作，用于根据游戏内资源和仓库存券计算当前和下一个卡池的抽数，并通过代理服务和 UI 对外提供。

New Features:
- 引入 PullCountCalculator 自定义动作，读取公招（招募）资源并扫描仓库存券，以计算当前和下一个招募池的可抽次数。
- 新增券配置数据、流水线、任务以及界面联接逻辑，使客户端 UI 可以触发抽数计算。

Enhancements:
- 为贵重物品仓库页面实现更健壮的基于 OCR 的扫描、页面去重与滚动检测，避免在扫描代金券时出现无限循环。

Tests:
- 添加单元测试，覆盖抽数公式、代金券汇总、OCR 整数解析、页面特征签名稳定性以及仓库网格变更检测等内容。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Add a PullCountCalculator custom action to compute current and next-banner pull counts from in-game resources and warehouse vouchers, and expose it through the agent service and UI.

New Features:
- Introduce a PullCountCalculator custom action that reads recruitment resources and scans warehouse vouchers to calculate pulls for the current and next recruitment pools.
- Add voucher configuration data, pipelines, tasks, and interface wiring to run the pull count calculation from the client UI.

Enhancements:
- Implement robust OCR-based scanning, page deduplication, and scroll detection for valuables warehouse pages to avoid infinite loops during voucher scans.

Tests:
- Add unit tests covering pull count formulas, voucher summarization, OCR integer parsing, page signature stability, and warehouse grid change detection.

</details>

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

添加一个 PullCountCalculator 自定义动作，用于根据游戏内资源和仓库存券计算当前和下一个卡池的抽数，并通过代理服务和 UI 对外提供。

New Features:
- 引入 PullCountCalculator 自定义动作，读取公招（招募）资源并扫描仓库存券，以计算当前和下一个招募池的可抽次数。
- 新增券配置数据、流水线、任务以及界面联接逻辑，使客户端 UI 可以触发抽数计算。

Enhancements:
- 为贵重物品仓库页面实现更健壮的基于 OCR 的扫描、页面去重与滚动检测，避免在扫描代金券时出现无限循环。

Tests:
- 添加单元测试，覆盖抽数公式、代金券汇总、OCR 整数解析、页面特征签名稳定性以及仓库网格变更检测等内容。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Add a PullCountCalculator custom action to compute current and next-banner pull counts from in-game resources and warehouse vouchers, and expose it through the agent service and UI.

New Features:
- Introduce a PullCountCalculator custom action that reads recruitment resources and scans warehouse vouchers to calculate pulls for the current and next recruitment pools.
- Add voucher configuration data, pipelines, tasks, and interface wiring to run the pull count calculation from the client UI.

Enhancements:
- Implement robust OCR-based scanning, page deduplication, and scroll detection for valuables warehouse pages to avoid infinite loops during voucher scans.

Tests:
- Add unit tests covering pull count formulas, voucher summarization, OCR integer parsing, page signature stability, and warehouse grid change detection.

</details>

</details>

## Summary by Sourcery

添加一个 PullCountCalculator 自定义动作，用于根据游戏内资源和仓库凭证计算当前与下一个卡池的抽数，并将其接入到代理服务和前端界面中。

New Features:
- 引入支持多阶段流水线的 PullCountCalculator 自定义动作，用于读取源石 / 合成玉等资源，并对仓库中的凭证进行分类，以计算可用抽数。
- 添加流水线、任务、数据结构(schema)和接口的连接配置以及多语言(i18n)文案，使客户端界面可以调用抽数计算功能。

Enhancements:
- 实现会话管理、OCR 整数提取、去重安全的凭证聚合以及详细日志记录，以提升抽数计算的健壮性和可调试性。

Tests:
- 添加 Go 单元测试，覆盖抽数公式、凭证累积去重逻辑以及 OCR 整数解析的边界情况。

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Add a PullCountCalculator custom action to compute current and next-banner pull counts from in-game resources and warehouse vouchers, and wire it into the agent service and UI.

New Features:
- Introduce a PullCountCalculator custom action with multi-stage pipeline support to read originium/oroberyl resources and classify warehouse vouchers for pull count calculation.
- Add pipeline, task, schema, and interface wiring plus i18n strings so the pull count calculation can be invoked from the client interface.

Enhancements:
- Implement session management, OCR integer extraction, duplicate-safe voucher aggregation, and detailed logging to make pull count computation robust and debuggable.

Tests:
- Add Go unit tests covering pull count formulas, voucher accumulation de-duplication, and OCR integer parsing edge cases.

</details>

### CODE REVIEW

**Comment by qiongtu2077** [comment]
@sourcery-ai 帮我再审核一下

File: assets/resource/pipeline/Interface/SceneMenu.json [hunk 1]
```diff
@@ -47,6 +47,7 @@
         "post_delay": 0,
         "rate_limit": 0,
         "next": [
+            "__ScenePrivateMenuListEnterMenuValuables",
             "__ScenePrivateWorldEnterMenuValuables",
```
Comment (zmdyy0318) [inline assets/resource/pipeline/Interface/SceneMenu.json:50]: __ScenePrivateWorldEnterMenuValuables里面是从MenuList进去的，这里加了一行是现有的代码有什么问题吗

File: agent/go-service/pullcount/action.go [hunk 2]
```diff
@@ -0,0 +1,1189 @@
...
+	return nil
+}
+
+// clickValuablesByVisibleText OCRs the right menu area and clicks the Valuables entry.
+func clickValuablesByVisibleText(ctx *maa.Context) error {
+	img, err := captureCurrentImage(ctx)
+	if err != nil {
+		return err
+	}
+	detail, err := ctx.RunRecognitionDirect(maa.RecognitionTypeOCR, maa.OCRParam{
```
Comment (zmdyy0318) [inline agent/go-service/pullcount/action.go]: 避免使用RunRecognitionDirect，尽量使用pipeline去做识别，否则跨平台适配会很麻烦

File: agent/go-service/pullcount/action.go [hunk 3]
```diff
@@ -0,0 +1,1189 @@
...
+		return err
+	}
+	if detail == nil {
+		return fmt.Errorf("%s returned nil detail", fallbackValuablesSwitch)
+	}
+	if !detail.Status.Success() {
+		return fmt.Errorf("%s status is %s", fallbackValuablesSwitch, detail.Status.String())
+	}
+	return nil
+}
```
Comment (zmdyy0318) [inline agent/go-service/pullcount/action.go]: 使用pipeline进行界面跳转，非必要不go

File: agent/go-service/pullcount/action.go [hunk 4]
```diff
@@ -0,0 +1,1189 @@
...
+			W: 58,
+			H: 32,
+		},
+		ScrollCheckROI: roiParam{
+			X: 40,
+			Y: 120,
+			W: 930,
+			H: 540,
+		},
+		ScrollChangeThreshold: 0.01,
```
Comment (zmdyy0318) [inline agent/go-service/pullcount/action.go]: roi写在pipeline里，非必要不go，否则跨平台适配很麻烦

File: agent/go-service/pullcount/action.go [hunk 5]
```diff
@@ -0,0 +1,1189 @@
...
+	controller.PostScreencap().Wait()
+	img, err := controller.CacheImage()
+	if err != nil {
+		return nil, err
+	}
+	if img == nil {
+		return nil, fmt.Errorf("cached image is nil")
+	}
+	return img, nil
+}
```
Comment (zmdyy0318) [inline agent/go-service/pullcount/action.go]: pipeline截图是有什么性能瓶颈吗，如果没有的话，go里避免使用截图

**Comment by zmdyy0318** [comment]
go太多了，一个算抽数并输出的action 1000多行go，应该用不了那么多

**Review by zmdyy0318 — CHANGES_REQUESTED** [review]

**Comment by ocsin1** [comment]
请查看相关[编码规范](https://github.com/MaaEnd/MaaEnd/blob/v2/docs/zh_cn/developers/coding-standards.md)

**Review by ocsin1 — CHANGES_REQUESTED** [review]
assets/resource/pipeline/PullCountCalculator.json这里面为什么要这样遍历,这会导致维护极度困难,复杂度比go还高,而且好像没有处理翻页?(

~~这种写法跟45个if有什么区别~~

**Comment by qiongtu2077** [comment]
处理翻页是有的：PullCountCalculatorWarehouseScrollDown。我发现PullCountCalculator.json不能把 45 格真正变成 Pipeline 里的 for，因为 MaaFramework Pipeline 没有原生 foreach/宏展开。官方协议里只有 next 顺序检测、repeat 重复同一个节点动作、max_hit、anchor、默认属性等能力。请问有没有什么好的解决方法呢@ocsin

**Comment by qiongtu2077** [comment]
如果只找券图标的话确实可以大幅缩减，但是需要特许寻访凭证图标、当前池限时券图标、寻访情报书图标。其中当前池限时券图标每个版本都不同需要手动加入。

**Comment by ocsin1** [comment]
> 处理翻页是有的：PullCountCalculatorWarehouseScrollDown。我发现PullCountCalculator.json不能把 45 格真正变成 Pipeline 里的 for，因为 MaaFramework Pipeline 没有原生 foreach/宏展开。官方协议里只有 next 顺序检测、repeat 重复同一个节点动作、max_hit、anchor、默认属性等能力。请问有没有什么好的解决方法呢@Ocsin

这里应该用 go ,我的思路是灰度处理后匹配分界线,切分不同的格子(形态学操作去掉噪声),每点击一次记录一下,然后重新识别再点击不重复的格子(记录格子的中心坐标),翻页动作之后对齐格子后重新执行上述(对齐和尾扫的代码应该可以参照基质识别)

纯pipeline也可以直接匹配45个格子然后用jumpback加"max": 1 实现foreach,可以参考环境监测的foreach实现方式(节点是靠游戏数据用工具生成的,复杂度不算很高,具体看环境监测的维护文档)

**Comment by qiongtu2077** [comment]
现在的流程是不做仓库全格遍历，而是基于 Pipeline 的模板匹配直接搜索目标券图标，抽数计算由 Pipeline 负责界面流程、OCR、模板匹配、点击与翻页，Go 只负责保存识别结果、去重、计算公式和输出日志。任务从寻访页读取嵌晶玉与衍质源石换算值，再跳转到贵重品库扫描可继承到下版本的特许寻访凭证和60抽免费券-“寻访请报书”，最后按“资源折算 + 可继承券 + 下版本商店 5 抽 + 签到 5 抽”输出当前池可用抽数和下版本池总抽数。

**Comment by zmdyy0318** [comment]
ai搞太多无意义代码了，估计后面出问题也只能让ai改，随后就是无意义封装指数级上升。
大概率是没想好go要怎么写，让ai自由发挥。

**Comment by zmdyy0318** [comment]
我有空了改一波

**Comment by zmdyy0318** [comment]
@copilot resolve the merge conflicts in this pull request
